go/pkg/client/cas.go

```diff
@@ -25,32 +26,131 @@ import (
-// Returns a slice of the missing digests.
-func (c *Client) UploadIfMissing(ctx context.Context, data ...*chunker.Chunker) ([]digest.Digest, error) {
-	if cap(c.casUploaders) <= 0 {
-		return nil, fmt.Errorf("CASConcurrency should be at least 1")
+type uploadRequest struct {
+	c    *chunker.Chunker
+	wait chan<- *uploadResponse
+}
+
+type uploadResponse struct {
+	digest  digest.Digest
+	err     error
+	missing bool
+}
+
+type uploadState struct {
+	data *chunker.Chunker
+	err  error
+
+	// mu protects clients. It does NOT protect data or error, because they do not need protection -
+	// they are only modified when a state object is created, and by updateAndNotify which is called
+	// exactly once for a given state object (this is the whole point of the algorithm).
+	mu      sync.Mutex
+	clients []chan<- *uploadResponse
+}
+
+func (c *Client) getMissingPresent(ctx context.Context, dgs []digest.Digest) (missing []digest.Digest, present []digest.Digest, err error) {
+	dgMap := make(map[digest.Digest]bool)
+	for _, d := range dgs {
+		dgMap[d] = true
+	}
+	missing, err = c.MissingBlobs(ctx, dgs)
+	for _, d := range missing {
+		delete(dgMap, d)
+	}
+	for d := range dgMap {
+		present = append(present, d)
+	}
+	return missing, present, err
+}
+
+func (c *Client) uploadProcessor() {
+	var buffer []*uploadRequest
+	ticker := time.NewTicker(100 * time.Millisecond)
+	for {
+		select {
+		case req, ok := <-c.casUploadRequests:
+			if !ok {
+				// Client is exiting. Notify remaining uploads to prevent deadlocks.
+				ticker.Stop()
+				if buffer != nil {
+					for _, r := range buffer {
+						r.wait <- &uploadResponse{err: context.Canceled}
+					}
+				}
+				return
+			}
+			buffer = append(buffer, req)
+			if len(buffer) >= casChanBufferSize {
+				c.upload(buffer)
+				buffer = nil
+			}
+		case <-ticker.C:
+			if buffer != nil {
+				c.upload(buffer)
+				buffer = nil
+			}
+		}
+	}
+}
+
+func updateAndNotify(st *uploadState, err error, missing bool) {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+	st.err = err
+	for i, cl := range st.clients {
+		cl <- &uploadResponse{
+			digest:  st.data.Digest(),
+			missing: i == 0 && missing, // Only first client is reported the digest missing.
+			err:     err,
+		}
 	}
+	st.clients = nil
+	st.data = nil
+}
+
+func (c *Client) upload(reqs []*uploadRequest) {
 	const (
 		logInterval = 25
 	)
 
-	var dgs []digest.Digest
-	chunkers := make(map[digest.Digest]*chunker.Chunker)
-	for _, c := range data {
-		dg := c.Digest()
-		if _, ok := chunkers[dg]; !ok {
-			dgs = append(dgs, dg)
-			chunkers[dg] = c
+	// Collect new uploads.
+	newStates := make(map[digest.Digest]*uploadState)
+	var newUploads []digest.Digest
+	for _, req := range reqs {
+		dg := req.c.Digest()
+		st, ok := c.casUploads[dg]
+		if ok {
+			st.mu.Lock()
+			if len(st.clients) > 0 {
+				st.clients = append(st.clients, req.wait)
+			} else {
+				req.wait <- &uploadResponse{err: nil, missing: false}
```

Comment:
IIUC, this is an error scenario? The digest exists in casUploaders, which means there should be some clients waiting on it but there aren't any. Should we log an error message if this happens?

Reply:
Thanks, I think I understand this a little better now. So the `upload()` method is called sequentially by `uploadProcessor()` and `upload()` spawns multiple go-routines that actually do the upload.

An `uploadState()` will exist in `casUploaders`, if we have already seen an upload request for the digest. Now coming to this specific if condition:

```
if len(st.clients) > 0 {
...
} else {
...
}
```
IIUC, the else part will happen when one of the go-routines spawned by `upload()` finished doing the upload (either successfully or unsuccessfully), acquired `st.mu` before the code on line `161` did and finished notifying all the clients.

So now my question is, if `len(st.clients) == 0`, then it just means that we notified all the clients of the result. How do we know whether the upload itself was successful or not? We seem to be returning `&uploadResponse{err: nil, missing: false}` which is assuming that we didn't run into an error with the upload?

Reply:
Correct, thank you, that was a mistake! I should have returned st.err. Done.